### DISCUSSION TITLE
Flutter drift web compilation failed - Target dart2js failed

### DISCUSSION THREAD

**Discussion by zalekbloom** [simolus3/drift#2131] [Q&A] [answered]
I though it as a flutter problem, so I posted this question on a flutter forums, they advice me it is a drift problem.
Following the advice I did:
flutter pub cache clean  
flutter clean
flutter pub get
flutter doctor –v   (result attached)
flutter build web

Result of flutter doctor –v 
flutter doctor -v
[√] Flutter (Channel stable, 3.3.6, on Microsoft Windows [Version 10.0.19044.2130], locale en-US)
    • Flutter version 3.3.6 on channel stable at c:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6928314d50 (11 days ago), 2022-10-25 16:34:41 -0400
    • Engine revision 3ad69d7be3
    • Dart version 2.18.2
    • DevTools version 2.15.0

[√] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at C:\Users\Yossi\AppData\Local\Android\sdk
    • Platform android-31, build-tools 32.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Build Tools 2019 16.11.10)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools
    • Visual Studio Build Tools 2019 version 16.11.32126.315
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2021.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19044.2130]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 107.0.5304.88
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 107.0.1418.35

[√] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

result of flutter build web
flutter build web

 Building with sound null safety 

Target dart2js failed: Exception: Warning: The 'dart2js' entrypoint script is deprecated, please use 'dart compile js' instead.
/C:/Users/Yossi/AppData/Local/Pub/Cache/git/sqflite_web-a5e7fd3e2e4b05f52728221e41752a318532f4a8/lib/src/sqflite_web_impl.dart:469:49:
Error: The parameter 'txn' of the method 'SqfliteWebDatabase.txnApplyBatch' has type 'SqfliteTransaction', which does not match the corresponding type,
'SqfliteTransaction?', in the overridden method, 'SqfliteDatabase.txnApplyBatch'.
 - 'SqfliteTransaction' is from 'package:sqflite_common/src/transaction.dart'
 ('/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.3.0/lib/src/transaction.dart').
  Future<List> txnApplyBatch(SqfliteTransaction txn, SqfliteBatch batch, {bool? noResult, bool? continueOnError}) {
                                                ^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.3.0/lib/src/database.dart:82:25:
Info: This is the overridden method ('txnApplyBatch').
  Future<List<Object?>> txnApplyBatch(
                        ^
Error: Compilation failed.

Compiling lib\main.dart for the Web...                             11.8s
Exception: Failed to compile application for the Web.

Any ideas what is the problem?

Thanks,

zb

**Comment by simolus3** [comment]
It looks like you got incompatible versions of `sqflite_web` and `sqflite_common` somehow. I'm not maintaining either of these packages, so I'm not sure how much can do about that. Did you check that you got the most recent version of both packages?

Are you using a `FlutterQueryExecutor`/`SqfliteExecutor` for the web? Using `sqflite_web` should work, but I've never tried it together with Drift. Drift also has [its own web support](https://drift.simonbinder.eu/web/) that you could try as well. A Flutter app running on all platforms with drift is available [in this repo](https://github.com/simolus3/drift/tree/develop/examples/app), maybe that helps to get Drift working for the web?

**Comment by zalekbloom** [comment, marked as answer]
I upgrader all my packages, currently command:  flutter pub outdated shows:

flutter pub outdated
Showing outdated packages.
[*] indicates versions that are not the latest available.

Package Name              Current   Upgradable  Resolvable  Latest

direct dependencies: all up-to-date.

dev_dependencies: all up-to-date.

transitive dependencies:
async                     *2.9.0    *2.9.0      *2.9.0      2.10.0
collection                *1.16.0   *1.16.0     *1.16.0     1.17.0
js                        *0.6.4    *0.6.4      *0.6.4      0.6.5
matcher                   *0.12.12  *0.12.12    *0.12.12    0.12.13
material_color_utilities  *0.1.5    *0.1.5      *0.1.5      0.2.0
source_span               *1.9.0    *1.9.0      *1.9.0      1.9.1
stack_trace               *1.10.0   *1.10.0     *1.10.0     1.11.0
stream_channel            *2.1.0    *2.1.0      *2.1.0      2.1.1
string_scanner            *1.1.1    *1.1.1      *1.1.1      1.2.0
vector_math               *2.1.2    *2.1.2      *2.1.2      2.1.4

transitive dev_dependencies:
boolean_selector          *2.1.0    *2.1.0      *2.1.0      2.1.1
test_api                  *0.4.12   *0.4.12     *0.4.12     0.4.16
all dependencies are up-to-date.

after flutter clean, flutter pub get, flutter pub cache clean   I am still getting:

flutter build web
Running "flutter pub get" in private_dictionary...                 12.6s

 Building with sound null safety

Target dart2js failed: Exception: Warning: The 'dart2js' entrypoint script is deprecated, please use 'dart compile js' instead.
/C:/Users/Yossi/AppData/Local/Pub/Cache/git/sqflite_web-a5e7fd3e2e4b05f52728221e41752a318532f4a8/lib/src/sqflite_web_impl.dart:80:7:
Error: The non-abstract class 'SqfliteWebDatabase' is missing implementations for these members:
 - DatabaseExecutor.queryCursor
 - DatabaseExecutor.rawQueryCursor
 - SqfliteDatabase.txnQueryCursorClose
 - SqfliteDatabase.txnQueryCursorGetCurrent
 - SqfliteDatabase.txnQueryCursorMoveNext
 - SqfliteDatabase.txnRawQueryCursor
class SqfliteWebDatabase extends SqfliteDatabase {
      ^^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/sqlite_api.dart:171:23:
Info: 'DatabaseExecutor.queryCursor' is defined here.
  Future<QueryCursor> queryCursor(String table,
                      ^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/sqlite_api.dart:166:23:
Info: 'DatabaseExecutor.rawQueryCursor' is defined here.
  Future<QueryCursor> rawQueryCursor(String sql, List<Object?>? arguments,
                      ^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:114:16:
Info: 'SqfliteDatabase.txnQueryCursorClose' is defined here.
  Future<void> txnQueryCursorClose(
               ^^^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:110:24:
Info: 'SqfliteDatabase.txnQueryCursorGetCurrent' is defined here.
  Map<String, Object?> txnQueryCursorGetCurrent(
                       ^^^^^^^^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:106:16:
Info: 'SqfliteDatabase.txnQueryCursorMoveNext' is defined here.
  Future<bool> txnQueryCursorMoveNext(
               ^^^^^^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:102:30:
Info: 'SqfliteDatabase.txnRawQueryCursor' is defined here.
  Future<SqfliteQueryCursor> txnRawQueryCursor(SqfliteTransaction? txn,
                             ^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/git/sqflite_web-a5e7fd3e2e4b05f52728221e41752a318532f4a8/lib/src/sqflite_web_impl.dart:469:49:
Error: The parameter 'txn' of the method 'SqfliteWebDatabase.txnApplyBatch' has type 'SqfliteTransaction', which does not match the corresponding type,
'SqfliteTransaction?', in the overridden method, 'SqfliteDatabase.txnApplyBatch'.
 - 'SqfliteTransaction' is from 'package:sqflite_common/src/transaction.dart'
 ('/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/transaction.dart').
  Future<List> txnApplyBatch(SqfliteTransaction txn, SqfliteBatch batch, {bool? noResult, bool? continueOnError}) {
                                                ^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:84:25:
Info: This is the overridden method ('txnApplyBatch').
  Future<List<Object?>> txnApplyBatch(
                        ^
/C:/Users/Yossi/AppData/Local/Pub/Cache/git/sqflite_web-a5e7fd3e2e4b05f52728221e41752a318532f4a8/lib/src/sqflite_web_impl.dart:500:13:
Error: The method 'SqfliteWebDatabase.txnExecute' has fewer named arguments than those of overridden method 'SqfliteDatabase.txnExecute'.
  Future<T> txnExecute<T>(SqfliteTransaction? txn, String sql, [List? arguments]) {
            ^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:89:13:
Info: This is the overridden method ('txnExecute').
  Future<T> txnExecute<T>(
            ^
Error: Compilation failed.


Compiling lib\main.dart for the Web...                             12.6s
Exception: Failed to compile application for the Web.

zb

________________________________

**Comment by zalekbloom** [comment]
I never placed a code with FlutterQueryExecutor/SqfliteExecutor, but when I tried to run my app in debug mode on Chrom, I got this error:

Launching lib\main.dart on Chrome in debug mode...
/C:/Users/Yossi/AppData/Local/Pub/Cache/git/sqflite_web-a5e7fd3e2e4b05f52728221e41752a318532f4a8/lib/src/sqflite_web_impl.dart:80:7: Error: The non-abstract class 'SqfliteWebDatabase' is missing implementations for these members:
 - DatabaseExecutor.queryCursor
 - DatabaseExecutor.rawQueryCursor
 - SqfliteDatabase.txnQueryCursorClose
 - SqfliteDatabase.txnQueryCursorGetCurrent
 - SqfliteDatabase.txnQueryCursorMoveNext
 - SqfliteDatabase.txnRawQueryCursor
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class SqfliteWebDatabase extends SqfliteDatabase {
      ^^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/sqlite_api.dart:171:23: Context: 'DatabaseExecutor.queryCursor' is defined here.

  Future<QueryCursor> queryCursor(String table,
                      ^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/sqlite_api.dart:166:23: Context: 'DatabaseExecutor.rawQueryCursor' is defined here.
  Future<QueryCursor> rawQueryCursor(String sql, List<Object?>? arguments,
                      ^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:114:16: Context: 'SqfliteDatabase.txnQueryCursorClose' is defined here.
  Future<void> txnQueryCursorClose(
               ^^^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:110:24: Context: 'SqfliteDatabase.txnQueryCursorGetCurrent' is defined here.
  Map<String, Object?> txnQueryCursorGetCurrent(
                       ^^^^^^^^^^^^^^^^^^^^^^^^

/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:106:16: Context: 'SqfliteDatabase.txnQueryCursorMoveNext' is defined here.
  Future<bool> txnQueryCursorMoveNext(
               ^^^^^^^^^^^^^^^^^^^^^^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:102:30: Context: 'SqfliteDatabase.txnRawQueryCursor' is defined here.
  Future<SqfliteQueryCursor> txnRawQueryCursor(SqfliteTransaction? txn,
                             ^^^^^^^^^^^^^^^^^

/C:/Users/Yossi/AppData/Local/Pub/Cache/git/sqflite_web-a5e7fd3e2e4b05f52728221e41752a318532f4a8/lib/src/sqflite_web_impl.dart:469:49: Error: The parameter 'txn' of the method 'SqfliteWebDatabase.txnApplyBatch' has type 'SqfliteTransaction', which does not match the corresponding type, 'SqfliteTransaction?', in the overridden method, 'SqfliteDatabase.txnApplyBatch'.
- 'SqfliteTransaction' is from 'package:sqflite_common/src/transaction.dart' ('/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/transaction.dart').
Change to a supertype of 'SqfliteTransaction?', or, for a covariant parameter, a subtype.
  Future<List> txnApplyBatch(SqfliteTransaction txn, SqfliteBatch batch, {bool? noResult, bool? continueOnError}) {

                                                ^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:84:25: Context: This is the overridden method ('txnApplyBatch').
  Future<List<Object?>> txnApplyBatch(
                        ^
/C:/Users/Yossi/AppData/Local/Pub/Cache/git/sqflite_web-a5e7fd3e2e4b05f52728221e41752a318532f4a8/lib/src/sqflite_web_impl.dart:500:13: Error: The method 'SqfliteWebDatabase.txnExecute' has fewer named arguments than those of overridden method 'SqfliteDatabase.txnExecute'.
  Future<T> txnExecute<T>(SqfliteTransaction? txn, String sql, [List? arguments]) {
            ^
/C:/Users/Yossi/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/sqflite_common-2.4.0+2/lib/src/database.dart:89:13: Context: This is the overridden method ('txnExecute').
  Future<T> txnExecute<T>(
            ^

Failed to compile application.
Exited (sigterm)


zb

________________________________

**Reply by simolus3** [reply]
If you're getting that error, you or one of your dependencies must import sqflite somehow. If you're not using `drift_sqflite`, it's probably another package. I can't help you with that in that case.
